### PR TITLE
(Closes #2253) allow for GH_WRITE access to continuous field args to inter-grid kernels

### LINKED ISSUE #2253: [LFRic] Error when using GH_WRITE and continuous fields in intergrid kernels
I've run into a very specific issue when using intergrid kernels which have a continuous field with the `GH_WRITE` permission, e.g.: https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/trunk/gungho/source/kernel/intermesh_transforms/restrict_w2_kernel_mod.F90?rev=41236

For kernels like this, we should loop up to `mesh%get_last_edge_cell()`, and there is generally no need to perform a halo exchanges.

However it seems that we are looping up to `mesh%get_last_halo_cell(1)`, and therefore may be performing some halo exchanges that we don't need to on the other fields. For intergrid kernels, we may be reading from a field on a fine mesh -- if we have more than 2 `ntarget_cells_per_source_x` or `ntarget_cells_per_source_y` then we may be trying to read from information that isn't in the halo cells of those fields, leading to a segmentation fault.

I can bypass this at the LFRic end by changing the kernel metadata so that it is `GH_WRITE` with some `ANY_DISCONTINUOUS_SPACE`.

This is very related to #1542, so happy to close this issue and just post an update on there.

### PR DESCRIPTION
**Pull request by arporter**

### CODE REVIEW

**Comment by arporter** [comment]
I'll set the integration tests running and then this just needs a bit of tidying.

**Comment by arporter** [comment]
Integration tests were all green.

**Comment by arporter** [comment]
On reflection I decided I could do better than simply declare both cell maps (and also, I think LFRic complains about unused variables). I've therefore tightened things up and have re-launched the integration tests.

**Comment by arporter** [comment]
This is now ready for testing by @thomasmelvin and @mo-lottieturner.

**Comment by arporter** [comment]
This has been tested by the MO and is now ready for review. One for @TeranIvy, @hiker or @rupertford.

**Review by TeranIvy — CHANGES_REQUESTED** [review]
Nice work! Tests with compilation pass, examples run, and the documentation builds fine. There are some minor comments (see inline).

From what I have gathered looking at the user guide, the behaviour for loop stops for MPI is not really explained (e.g. looping into halos vs looping until edge cells regarding the function space continuity). Now, with GH_WRITE thrown into the mix for continuous function spaces, things can get confusing, as we still need to apply colouring but we have loop limits for discontinuous function spaces. It would be good to clarify what PSyclone generates.

**Comment by arporter** [comment]
> From what I have gathered looking at the user guide, the behaviour for loop stops for MPI is not really explained (e.g. looping into halos vs looping until edge cells regarding the function space continuity). Now, with GH_WRITE thrown into the mix for continuous function spaces, things can get confusing, as we still need to apply colouring but we have loop limits for discontinuous function spaces. It would be good to clarify what PSyclone generates.

This is covered in the developer guide: https://psyclone-dev.readthedocs.io/en/latest/APIs.html#cell-iterators-continuous

File: src/psyclone/tests/dynamo0p3_multigrid_test.py [hunk 1]
```diff
@@ -507,6 +506,30 @@ def test_field_restrict(tmpdir, dist_mem, monkeypatch, annexed):
         assert set_dirty in output
 
 
+def test_cont_field_restrict(tmpdir, dist_mem, monkeypatch, annexed):
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/dynamo0p3_multigrid_test.py:509]: Oops. Added missing compilation check.

File: src/psyclone/tests/dynamo0p3_multigrid_test.py [hunk 2]
```diff
@@ -507,6 +506,30 @@ def test_field_restrict(tmpdir, dist_mem, monkeypatch, annexed):
...
+    ''' Test that we generate correct code for an invoke containing a
+    single restriction operation (read from fine, write to
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/dynamo0p3_multigrid_test.py:511]: Oh yes, that's clearer.

File: src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py [hunk 3]
```diff
@@ -414,6 +414,35 @@ def test_colour_trans_continuous_write(dist_mem, tmpdir):
     assert LFRicBuild(tmpdir).code_compiles(psy)
 
 
+def test_colour_continuous_writer_intergrid(tmpdir, dist_mem):
+    '''Test the loop-colouring transformation for an intergrid kernel that has
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/domain/lfric/transformations/dynamo0p3_transformations_test.py:418]: Done (albeit new-line before Test as I prefer that). The dev guide documents the need to loop into the halos.

File: src/psyclone/tests/test_files/dynamo0p3/restrict_w2_kernel_mod.F90 [hunk 4]
```diff
@@ -0,0 +1,388 @@
+!-----------------------------------------------------------------------------
+! (C) Crown copyright 2021 Met Office. All rights reserved.
+! The file LICENCE, distributed with this code, contains details of the terms
+! under which the code may be used.
+!-----------------------------------------------------------------------------
+! LICENCE.original is available from the Met Office Science Repository Service:
+! https://code.metoffice.gov.uk/trac/lfric/browser/LFRic/trunk/LICENCE.original
+! -----------------------------------------------------------------------------
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/restrict_w2_kernel_mod.F90:8]: Added.

File: src/psyclone/tests/test_files/dynamo0p3/restrict_w2_kernel_mod.F90 [hunk 5]
```diff
@@ -0,0 +1,388 @@
...
+
+implicit none
+
+private
+
+!-------------------------------------------------------------------------------
+! Public types
+!-------------------------------------------------------------------------------
+!> The type declaration for the kernel. Contains the metadata needed by the
+!> Psy layer.
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/tests/test_files/dynamo0p3/restrict_w2_kernel_mod.F90:36]: Done.

**Comment by arporter** [comment]
CI-permitting, this is ready for another look. I've also fired off the integration tests.

**Review by TeranIvy — COMMENTED** [review]
Thanks for making changes. All looks good now, so I am proceeding to final testing.

**Review by TeranIvy — APPROVED** [review]
CI and tests with compilation are fine. Updated changelog and UG for merge to master.

**Pull request merged**